src/platforms/hosted/libftdi_jtagtap.c

```diff
@@ -28,98 +28,120 @@
+	libftdi_drain_potential_garbage();
+
+	const uint8_t cmd[6] = {
+		SET_BITS_LOW,
+		active_state.data_low,
+		active_state.ddr_low,
+		SET_BITS_HIGH,
+		active_state.data_high,
+		active_state.ddr_high,
+	};
+	libftdi_buffer_write_arr(cmd);
 	libftdi_buffer_flush();
 	/* Write out start condition and pull garbage from read buffer.
 	 * FT2232D otherwise misbehaves on runs follwoing the first run.*/
-	garbage = ftdi_read_data(ftdic, cmd_write, sizeof(cmd_write));
-	if (garbage > 0) {
-		DEBUG_WARN("FTDI JTAG end init got garbage:");
-		for (int i = 0; i < garbage; i++)
-			DEBUG_WARN(" %02x", cmd_write[i]);
-		DEBUG_WARN("\n");
-	}
-	/* Go to JTAG mode for SWJ-DP */
-	for (int i = 0; i <= 50; i++)
-		jtag_proc->jtagtap_next(1, 0);      /* Reset SW-DP */
-	jtag_proc->jtagtap_tms_seq(0xE73C, 16); /* SWD to JTAG sequence */
-	jtag_proc->jtagtap_tms_seq(0x1F, 6);
+	libftdi_drain_potential_garbage();
 
-	return 0;
+	/* Ensure we're in JTAG mode */
+	for (size_t i = 0; i <= 50; ++i)
+		jtagtap_next(true, false); /* 50 idle cycles for SWD reset */
+	jtagtap_tms_seq(0xe73cU, 16);  /* SWD to JTAG sequence */
+	jtagtap_soft_reset();
+	return true;
 }
 
 static void jtagtap_reset(void)
 {
 	jtagtap_soft_reset();
 }
 
-static void jtagtap_tms_seq(uint32_t tms_states, size_t ticks)
+static void jtagtap_tms_seq(uint32_t tms_states, const size_t clock_cycles)
 {
-	uint8_t tmp[3] = {MPSSE_WRITE_TMS | MPSSE_LSB | MPSSE_BITMODE | MPSSE_WRITE_NEG, 0, 0};
-	while (ticks > 0) {
-		tmp[1] = ticks < 7 ? ticks - 1 : 6;
-		tmp[2] = 0x80 | (tms_states & 0x7F);
-
-		libftdi_buffer_write(tmp, 3);
-		tms_states >>= 7;
-		if (ticks < 7)
-			ticks = 0;
-		else
-			ticks -= 7;
+	for (size_t cycle = 0; cycle < clock_cycles; cycle += 7) {
+		const uint8_t cmd[3] = {
+			MPSSE_WRITE_TMS | MPSSE_LSB | MPSSE_BITMODE | MPSSE_WRITE_NEG,
+			MIN(7U, clock_cycles - cycle) - 1U,
+			0x80U | (tms_states & 0x7FU),
+		};
+		tms_states >>= 7U;
+		libftdi_buffer_write_arr(cmd);
 	}
 }
 
-static void jtagtap_tdi_seq(const bool final_tms, const uint8_t *DI, size_t ticks)
+static void jtagtap_tdi_seq(const bool final_tms, const uint8_t *const data_in, const size_t clock_cycles)
 {
-	return libftdi_jtagtap_tdi_tdo_seq(NULL, final_tms, DI, ticks);
+	return libftdi_jtagtap_tdi_tdo_seq(NULL, final_tms, data_in, clock_cycles);
 }
 
-static bool jtagtap_next(bool dTMS, bool dTDI)
+static bool jtagtap_next(bool tms, bool tdi)
 {
-	uint8_t ret;
-	uint8_t tmp[3] = {MPSSE_WRITE_TMS | MPSSE_DO_READ | MPSSE_LSB | MPSSE_BITMODE | MPSSE_WRITE_NEG, 0, 0};
-	tmp[2] = (dTDI ? 0x80 : 0) | (dTMS ? 0x01 : 0);
-	libftdi_buffer_write(tmp, 3);
-	libftdi_buffer_read(&ret, 1);
-
-	ret &= 0x80;
+	const uint8_t cmd[3] = {
+		MPSSE_WRITE_TMS | MPSSE_DO_READ | MPSSE_LSB | MPSSE_BITMODE | MPSSE_WRITE_NEG,
+		0,
+		(tdi ? 0x80U : 0U) | (tms ? 0x01U : 0U),
+	};
+	libftdi_buffer_write_arr(cmd);
 
-	return ret;
+	uint8_t ret = 0;
+	libftdi_buffer_read_val(ret);
+	return ret & 0x80U;
```

Comment:
help me out with these functions, they come from the ftdi lib?
`libftdi_buffer_write_arr` doesn't take size arguments?

doesn't `libftdi_buffer_read_val` expect a pointer? or return the data? looks like we are checking for (0 & 0x80)

Reply:
libftdi_buffer_write_arr() and the other functions based on the names `libftdi_buffer_write` and `libftdi_buffer_read` (but not exact matches) are actually macros defined in the ftdi_bmp.h header:

https://github.com/blackmagic-debug/blackmagic/blob/f4d6746/src/platforms/hosted/ftdi_bmp.h#L102-L104

Reply:
ah! I did a find through github but it didn't find any

Reply:
ok, looking at it it looks correct, though I feel the macros just obfuscate things?

Reply:
they probably prevent mistakes

Reply:
They're designed to automate the size parameters to reduce mistakes and make it more clear that "this is reading an array", "this is writing an array", etc.. C doesn't provide us a better solution for this

Reply:
sure, that is a good reason for them, just when looking through with the C lenses on it's not immediately obvious